### PR TITLE
Feature/set up claim contract

### PR DESCRIPTION
**Pull request by VargaElod23**
# Generic Pull Request template

## Linked issue

Please link the issue:

ex. Closes #<33>.

## Checklist before requesting a review

- [ ] I have performed a self-review of my code
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics?
- [ ] Will this be part of a product update? If yes, please write one phrase about this update.
- [ ] The changes impact front-end.

## For changes impacting frontend

Please leave some screenshots/video recordings of the newly implemented interface and the Figma link for its initial design.

### Screenshots/ recordings

=> Multimedia here.

### Figma link

=> Figma link here.

## Description

Please provide a short description about what the PR contains.

## Funny gif

Leave a funny gif for faster review.

### CODE REVIEW

File: packages/contracts/contracts/DynamicEscrow.sol [hunk 1]
```diff
@@ -148,7 +101,7 @@ contract DynamicEscrow is IEscrow, Ownable, Pausable, ReentrancyGuard {
             ERC20 token = ERC20(tokenAddress);
             uint256 balance = token.balanceOf(spender);
             require(balance >= amount, "Insufficient balance");
-            token.transferFrom(spender, address(this), amount);
+            token.transferFrom(spender, _trustedAccountAddress, amount);
```
Comment (VargaElod23) [inline packages/contracts/contracts/DynamicEscrow.sol:104]: Keep funds in this contract

File: packages/contracts/contracts/DynamicEscrow.sol [hunk 2]
```diff
@@ -157,6 +110,37 @@ contract DynamicEscrow is IEscrow, Ownable, Pausable, ReentrancyGuard {
...
+    ) external override nonReentrant {
+        bytes32 hash = _hash(receiver, amount, nonce);
+
+        require(ECDSA.recover(hash, sig) == _trustedAccountAddress, "Claim: Invalid signature");
+
+        if (tokenAddress == address(0)) {
+            receiver.transfer(amount);
+        } else {
+            ERC20 token = ERC20(tokenAddress);
+            token.transferFrom(_trustedAccountAddress, receiver, amount);
```
Comment (VargaElod23) [inline packages/contracts/contracts/DynamicEscrow.sol:139]: Transfer from this address

File: packages/contracts/contracts/DynamicEscrow.sol [hunk 3]
```diff
@@ -171,20 +155,30 @@ contract DynamicEscrow is IEscrow, Ownable, Pausable, ReentrancyGuard {
...
             _deposits[poolId][msg.sender].weiAmount -= amount;
         }
-        if (_deposits[poolId][msg.sender].tokenAddress == address(0)) {
+        if (tokenAddress == address(0)) {
             receiver.transfer(amount);
         } else {
-            ERC20 token = ERC20(_deposits[poolId][msg.sender].tokenAddress);
-            token.transferFrom(address(this), receiver, amount);
+            ERC20 token = ERC20(tokenAddress);
+            token.transferFrom(_trustedAccountAddress, receiver, amount);
```
Comment (VargaElod23) [inline packages/contracts/contracts/DynamicEscrow.sol:169]: Add a test scenario where the receiver is a different address.

Comment (VargaElod23) [reply packages/contracts/contracts/DynamicEscrow.sol]: ERC20 test: https://github.com/Taraxa-project/hype/pull/168/files#diff-1525488124055a6eaccf6d76e23ddf74dc1cc61bd1eebd034e6adaf461cdbaaaR700
ETH test: https://github.com/Taraxa-project/hype/pull/168/files#diff-1525488124055a6eaccf6d76e23ddf74dc1cc61bd1eebd034e6adaf461cdbaaaR376

Comment (VargaElod23) [reply packages/contracts/contracts/DynamicEscrow.sol]: @rattrap  Tests for this scenario above

**Review by Malak67 — APPROVED** [review]

**Pull request merged**